core/module.go

```diff
@@ -4,66 +4,568 @@ import (
+			}
+			if ok {
+				return depType, true, nil
+			}
+		}
+
+		var found bool
+		// otherwise it must be from this module
+		for _, obj := range mod.ObjectDefs {
+			if obj.AsObject.Value.Name == typeDef.AsObject.Value.Name {
+				modType = &ModuleObjectType{
+					typeDef: obj.AsObject.Value,
+					mod:     mod,
+				}
+				found = true
+				break
+			}
+		}
+		if !found {
+			slog.Debug("module did not find object", "mod", mod.Name(), "object", typeDef.AsObject.Value.Name)
+			return nil, false, nil
+		}
+
+	case TypeDefKindInterface:
+		if checkDirectDeps {
+			// check to see if this is from a *direct* dependency
+			depType, ok, err := mod.Deps.ModTypeFor(ctx, typeDef)
+			if err != nil {
+				return nil, false, fmt.Errorf("failed to get interface type from dependency: %w", err)
+			}
+			if ok {
+				return depType, true, nil
+			}
+		}
+
+		var found bool
+		// otherwise it must be from this module
+		for _, iface := range mod.InterfaceDefs {
+			if iface.AsInterface.Value.Name == typeDef.AsInterface.Value.Name {
+				modType = &InterfaceType{
+					mod:     mod,
+					typeDef: iface.AsInterface.Value,
+				}
+				found = true
+				break
+			}
+		}
+		if !found {
+			slog.Debug("module did not find interface", "mod", mod.Name(), "object", typeDef.AsInterface.Value.Name)
+			return nil, false, nil
+		}
+
+	default:
+		return nil, false, fmt.Errorf("unexpected type def kind %s", typeDef.Kind)
+	}
+
+	if typeDef.Optional {
+		modType = &NullableType{
+			Elem:       typeDef.WithOptional(false),
+			Underlying: modType,
+		}
+	}
+
+	return modType, true, nil
+}
+
+// verify the typedef is has no reserved names
+func (mod *Module) validateTypeDef(ctx context.Context, typeDef *TypeDef) error {
+	switch typeDef.Kind {
+	case TypeDefKindList:
+		return mod.validateTypeDef(ctx, typeDef.AsList.Value.ElementTypeDef)
+	case TypeDefKindObject:
+		return mod.validateObjectTypeDef(ctx, typeDef)
+	case TypeDefKindInterface:
+		return mod.validateInterfaceTypeDef(ctx, typeDef)
+	}
+	return nil
+}
+
+func (mod *Module) validateObjectTypeDef(ctx context.Context, typeDef *TypeDef) error {
+	// check whether this is a pre-existing object from core or another module
+	modType, ok, err := mod.Deps.ModTypeFor(ctx, typeDef)
+	if err != nil {
+		return fmt.Errorf("failed to get mod type for type def: %w", err)
+	}
+	if ok {
+		if sourceMod := modType.SourceMod(); sourceMod != nil && sourceMod != mod {
+			// already validated, skip
+			return nil
+		}
+	}
+
+	obj := typeDef.AsObject.Value
+
+	for _, field := range obj.Fields {
+		if gqlFieldName(field.Name) == "id" {
+			return fmt.Errorf("cannot define field with reserved name %q on object %q", field.Name, obj.Name)
+		}
+		fieldType, ok, err := mod.Deps.ModTypeFor(ctx, field.TypeDef)
+		if err != nil {
+			return fmt.Errorf("failed to get mod type for type def: %w", err)
+		}
+		if ok {
+			if sourceMod := fieldType.SourceMod(); sourceMod != nil && sourceMod.Name() != ModuleName && sourceMod != mod {
```

Comment:
Checking both the name and the pointer (?) equality looks a little weird, not sure if typo or purposeful

Reply:
I thiiiink this is right, and the same thing it did before. It's basically trying to say "if the source module is the core module, that's OK, but if it's any module but myself, that's not OK" - since we allow e.g. `*Container` but require local interfaces for anything else.

Reply:
Oooh I misread `ModuleName`, didn't realize it was the core mod name. Make sense 👍

As a nit: I just realized the `// already validated, skip` comment right below this line is probably out of place.